### ISSUE TITLE
Unable to resolve unexported referenced type when moved

### ISSUE THREAD

**Issue by njlynch** [aws/jsii#1980]
### Affected Languages
- [X] `TypeScript` or `Javascript`
- [ ] `Python`
- [ ] `Java`
- [ ] .NET (`C#`, `F#`, ...)

### General Information
* **JSII Version:** 1.12.0
* **Platform:** Darwin 38f9d3d7d3a1.ant.amazon.com 18.7.0 Darwin Kernel Version 18.7.0: Thu Jun 18 20:50:10 PDT 2020; root:xnu-4903.278.43~1/RELEASE_X86_64 x86_64

### What is the problem?
When refactoring unexported (supporting) types from inline in the same file as exported types to another (unexported) file, jsii loses the ability to resolve the types.

#### This works

`index.ts`
```ts
export * from './has-unexported-base';
```

`has-unexported-base.ts`
```ts
interface MyExportedPropsBase {  // Not exported
  readonly myProperty: number;
}
export interface MyExportedProps extends MyExportedPropsBase { // Exported
  readonly mySecondProperty: number;
}
```

#### This doesn't

`index.ts`
```ts
export * from './has-unexported-base';
```

`has-unexported-base.ts`
```ts
import { MyExportedPropsBase } from './private/unexported-base';
export interface MyExportedProps extends MyExportedPropsBase {
  readonly mySecondProperty: number;
}
```

`private/unexported-base.ts`
```ts
export interface MyExportedPropsBase {  // Exported here, but not in `index.ts`
  readonly myProperty: number;
}
```

### Verbose Log

```
[2020-09-04T11:58:30.924] [ERROR] jsii/compiler - Type model errors prevented the JSII assembly from being created
error JSII9002: Unable to resolve type "jsii-calc.MyExportedPropsBase". It may be @iternal or not exported from the module's entry point (as configured in "package.json" as "main").
lib/has-unexported-base.ts:3:1 - error JSII9002: Unable to resolve type "jsii-calc.MyExportedPropsBase". It may be @iternal or not exported from the module's entry point (as configured in "package.json" as "main").

3 export interface MyExportedProps extends MyExportedPropsBase {
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4   readonly mySecondProperty: number;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
5 }
```

**Comment by njlynch** [comment]
Also, just noticed the typo in the error message above:
`It may be @iternal or not exported...`

https://github.com/aws/jsii/blob/master/packages/jsii/lib/jsii-diagnostic.ts#L710

Worth fixing that at the same time? :D

**Comment by RomainMuller** [comment]
Your un-exported (from the package) type is exported from the module (i.e: file) that defines it. It must be annotated with `@internal` as you are now required to demonstrate that you are *intentionally* not forwarding it all the way to the root.

I guess it'd be sweet if we could have a warning that this non-`@internal` type isn't exported from the root, and that if this was intentional it should be annotated with `@internal`.

**Comment by njlynch** [comment]
Thanks for the clarification; I agree a more explicit warning in this case would be really useful.

**Comment by rix0rrr** [comment]
> I guess it'd be sweet if we could have a warning that this non-`@internal` type isn't exported from the root, and that if this was intentional it should be annotated with `@internal`.

If we're going to that I'd like to combine this with other heuristics: that a type in a `/private/` folder or with `_underscore.ts` filename should not be exported (and error if it is).

**Comment by njlynch** [comment]
One snag with the proposed workaround -- other consuming modules see the imports for the unexported classes/interfaces in the `*.d.ts` files, and fail to build.

See https://github.com/aws/aws-cdk/pull/10338 and the subsequent [build failure](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=20436d61-fa2b-4f12-bb22-d56af0e097bd%2Fbuild.log).

<details><summary>Build log failure</summary>

```
@aws-cdk/aws-backup: $ cdk-build
@aws-cdk/aws-backup: [2020-09-14T11:42:48.164] [ERROR] jsii/compiler - Compilation errors prevented the JSII assembly from being created
@aws-cdk/aws-backup: ../aws-rds/lib/cluster.d.ts:7:10 - error TS2305: Module '"./private/cluster-internal"' has no exported member 'DatabaseClusterBaseProps'.
@aws-cdk/aws-backup: 7 import { DatabaseClusterBaseProps, DatabaseClusterNew } from './private/cluster-internal';
@aws-cdk/aws-backup:            ~~~~~~~~~~~~~~~~~~~~~~~~
@aws-cdk/aws-backup: ../aws-rds/lib/cluster.d.ts:7:36 - error TS2724: Module '"./private/cluster-internal"' has no exported member 'DatabaseClusterNew'. Did you mean 'DatabaseClusterBase'?
@aws-cdk/aws-backup: 7 import { DatabaseClusterBaseProps, DatabaseClusterNew } from './private/cluster-internal';
@aws-cdk/aws-backup:                                      ~~~~~~~~~~~~~~~~~~
@aws-cdk/aws-backup: Error: /Users/nlynch/cdk/aws-cdk/node_modules/jsii/bin/jsii --silence-warnings=reserved-word exited with error code 1
@aws-cdk/aws-backup: Build failed. Total time (5.4s) | /Users/nlynch/cdk/aws-cdk/node_modules/jsii/bin/jsii (5.2s) | cfn2ts (0.2s)
```

</details>

Note that the above build failures are for the `aws-rds` `cluster.d.ts` file, referenced from the `aws-backup` module. The `aws-rds` module itself builds, but can't be used by the other consuming modules.

**Comment by RomainMuller** [comment]
Yeah we need to *not* erase `@internal` from the `.d.ts` files (this is achieved by changing a typescript compiler option).

> @rix0rrr: If we're going to that I'd like to combine this with other heuristics: that a type in a /private/ folder or with _underscore.ts filename should not be exported (and error if it is).

This would be a breaking change; and this looks too much like opinions forced on developers, which I'm not too favorable to.

**Comment by nikovirtala** [comment]
I tried to update one of my constructs to cdk v2 and apparently hit this issue(?). Is there anything I can do to work around this or is it even something I can workaround as a library user?

You can find full details of the error in this [gist](https://gist.github.com/nikovirtala/b165ff5652fd60410706afe786a2b2c2).

**Comment by nlongton** [comment]
I'm also getting this - i have a method that returns a CDK API Gateway class and JSII 

`
import {Annotations, CfnOutput, IAspect, Stack} from 'aws-cdk-lib';
import {aws_apigateway as api} from 'aws-cdk-lib';
import {aws_iam as iam} from 'aws-cdk-lib';
import {IConstruct} from 'constructs';

export class ApiGatewayHelpers {
  public static endpointConfiguration(stack: Stack, group?: string): api.EndpointConfiguration {
    const vpcInfo = VpcHelpers.lookupVpcInfo(Stack.of(stack), group)
    return {types: [api.EndpointType.PRIVATE], vpcEndpoints: [vpcInfo.apigwEndpointId]}
  }
`

error JSII3002: Type "aws-cdk-lib.EndpointConfiguration" cannot be used as a return type because it is not exported from aws-cdk-lib

9   public static endpointConfiguration(stack: Stack, group?: string): api.EndpointConfiguration {
                  ~~~~~~~~~~~~~~~~~~~~~

  node_modules/aws-cdk-lib/aws-apigateway/lib/restapi.d.ts:626:1
    626 export interface EndpointConfiguration {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    627     /**
        ~~~~~~~
    ...
    640     readonly vpcEndpoints?: IVpcEndpoint[];
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    641 }
        ~
    The referenced type is declared here
lib/api-gateway-helpers.ts:44:17 - error JSII3002: Type "aws-cdk-lib.PolicyDocument" cannot be used as a return type because it is not exported from aws-cdk-lib

**Comment by njlynch** [comment]
@nlongton - It sounds like you may be running into https://github.com/aws/aws-cdk/issues/17860. Can you leave a comment there as well?
